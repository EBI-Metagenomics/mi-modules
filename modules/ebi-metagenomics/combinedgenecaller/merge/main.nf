process COMBINEDGENECALLER_MERGE {
    tag "${meta.id}"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/mgnify-pipelines-toolkit:0.2.1--pyhdfd78af_0':
        'biocontainers/mgnify-pipelines-toolkit:0.2.1--pyhdfd78af_0' }"

    input:
    tuple val(meta), path(prodigal_sco, stageAs: "prodigal/"), path(prodigal_ffn, stageAs: "prodigal/"), path(prodigal_faa, stageAs: "prodigal/"), path(fgs_out, stageAs: "fgsrs/"), path(fgs_ffn, stageAs: "fgsrs/"), path(fgs_faa, stageAs: "fgsrs/"), path(mask)

    output:
    tuple val(meta), path("*.faa.gz"), emit: faa
    tuple val(meta), path("*.ffn.gz"), emit: ffn
    tuple val(meta), path("*.out.gz"), emit: summary
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    // Check if files are compressed
    def is_prod_sco_compressed = prodigal_sco.name.endsWith(".gz")
    def is_prod_ffn_compressed = prodigal_ffn.name.endsWith(".gz")
    def is_prod_faa_compressed = prodigal_faa.name.endsWith(".gz")
    def is_fgs_out_compressed = fgs_out.name.endsWith(".gz")
    def is_fgs_ffn_compressed = fgs_ffn.name.endsWith(".gz")
    def is_fgs_faa_compressed = fgs_faa.name.endsWith(".gz")

    // Get uncompressed file names
    def prodigal_sco_file = prodigal_sco.name.replace(".gz", "")
    def prodigal_ffn_file = prodigal_ffn.name.replace(".gz", "")
    def prodigal_faa_file = prodigal_faa.name.replace(".gz", "")
    def fgs_out_file = fgs_out.name.replace(".gz", "")
    def fgs_ffn_file = fgs_ffn.name.replace(".gz", "")
    def fgs_faa_file = fgs_faa.name.replace(".gz", "")

    def is_mask_compressed = false
    def mask_parameter = ""
    if ( mask ) {
        is_mask_compressed = mask.name.endsWith(".gz")
        mask_parameter = "--mask ${mask.name.replace(".gz", "")} "
    }
    """
    if [ "${is_prod_sco_compressed}" == "true" ]; then
        gzip -d -c ${prodigal_sco} > ${prodigal_sco_file}
    fi
    if [ "${is_prod_ffn_compressed}" == "true" ]; then
        gzip -d -c ${prodigal_ffn} > ${prodigal_ffn_file}
    fi
    if [ "${is_prod_faa_compressed}" == "true" ]; then
        gzip -d -c ${prodigal_faa} > ${prodigal_faa_file}
    fi
    if [ "${is_fgs_out_compressed}" == "true" ]; then
        gzip -d -c ${fgs_out} > ${fgs_out_file}
    fi
    if [ "${is_fgs_ffn_compressed}" == "true" ]; then
        gzip -d -c ${fgs_ffn} > ${fgs_ffn_file}
    fi
    if [ "${is_fgs_faa_compressed}" == "true" ]; then
        gzip -d -c ${fgs_faa} > ${fgs_faa_file}
    fi
    if [[ "${is_mask_compressed}" == "true" ]]; then
        gunzip ${mask}
    fi

    cgc_merge \\
        ${args} \\
        -n ${prefix} \\
        --prodigal-out ${prodigal_sco_file} \\
        --prodigal-ffn ${prodigal_ffn_file} \\
        --prodigal-faa ${prodigal_faa_file} \\
        --fgs-out ${fgs_out_file} \\
        --fgs-ffn ${fgs_ffn_file} \\
        --fgs-faa ${fgs_faa_file} ${mask_parameter}

    gzip -n ${prefix}.{faa,ffn,out}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
         mgnify-pipelines-toolkit: \$(get_mpt_version)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.{faa,ffn,out}

    gzip -n ${prefix}.{faa,ffn,out}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
         mgnify-pipelines-toolkit: \$(get_mpt_version)
    END_VERSIONS
    """
}
