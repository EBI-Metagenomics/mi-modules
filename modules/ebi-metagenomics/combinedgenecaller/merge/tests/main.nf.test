nextflow_process {

    name "Test Process COMBINEDGENECALLER_MERGE"
    script "../main.nf"
    process "COMBINEDGENECALLER_MERGE"

    tag "modules"
    tag "modules_nfcore"
    tag "modules_ebimetagenomics"
    tag "combinedgenecaller"
    tag "combinedgenecaller/merge"

    test("combinedgenecaller merge without a maskfile") {

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ],
                    file('${moduleDir}/tests/fixtures/input_prodigal.sco.gz', checkIfExists: true),
                    file('${moduleDir}/tests/fixtures/input_prodigal.ffn.gz', checkIfExists: true),
                    file('${moduleDir}/tests/fixtures/input_prodigal.faa.gz', checkIfExists: true),
                    file('${moduleDir}/tests/fixtures/input_fgs.out.gz', checkIfExists: true),
                    file('${moduleDir}/tests/fixtures/input_fgs.ffn.gz', checkIfExists: true),
                    file('${moduleDir}/tests/fixtures/input_fgs.faa.gz', checkIfExists: true),
                    []
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("combinedgenecaller merge with a maskfile") {

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file('${moduleDir}/tests/fixtures/input_prodigal.sco.gz', checkIfExists: true),
                    file('${moduleDir}/tests/fixtures/input_prodigal.ffn.gz', checkIfExists: true),
                    file('${moduleDir}/tests/fixtures/input_prodigal.faa.gz', checkIfExists: true),
                    file('${moduleDir}/tests/fixtures/input_fgs.out.gz', checkIfExists: true),
                    file('${moduleDir}/tests/fixtures/input_fgs.ffn.gz', checkIfExists: true),
                    file('${moduleDir}/tests/fixtures/input_fgs.faa.gz', checkIfExists: true),
                    file('${moduleDir}/tests/fixtures/cmsearch.all.tblout.deoverlapped.gz', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
