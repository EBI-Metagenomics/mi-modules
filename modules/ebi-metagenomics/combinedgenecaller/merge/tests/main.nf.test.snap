{
    "combinedgenecaller merge without a maskfile": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.faa.gz:md5,7ce6ef4028ea567c8ba920d26de3ef65"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.ffn.gz:md5,2984c9d227d6a91bfebdcf9709c16a54"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.out.gz:md5,df20ed72adc003d1e6b53ed8135c4b94"
                    ]
                ],
                "3": [
                    "versions.yml:md5,47831ed291d8ee32f13260f28865f68e"
                ],
                "faa": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.faa.gz:md5,7ce6ef4028ea567c8ba920d26de3ef65"
                    ]
                ],
                "ffn": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.ffn.gz:md5,2984c9d227d6a91bfebdcf9709c16a54"
                    ]
                ],
                "summary": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.out.gz:md5,df20ed72adc003d1e6b53ed8135c4b94"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,47831ed291d8ee32f13260f28865f68e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-12T14:43:01.553401"
    },
    "combinedgenecaller merge with a maskfile": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.faa.gz:md5,1577a0613f7fa5c1d989dd5bd2b81397"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.ffn.gz:md5,9bac7cc2f074517fd8af849f0f3839e6"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.out.gz:md5,bf80a3415d4b404716f88b16f75636d9"
                    ]
                ],
                "3": [
                    "versions.yml:md5,47831ed291d8ee32f13260f28865f68e"
                ],
                "faa": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.faa.gz:md5,1577a0613f7fa5c1d989dd5bd2b81397"
                    ]
                ],
                "ffn": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.ffn.gz:md5,9bac7cc2f074517fd8af849f0f3839e6"
                    ]
                ],
                "summary": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.out.gz:md5,bf80a3415d4b404716f88b16f75636d9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,47831ed291d8ee32f13260f28865f68e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-12T14:43:08.520856"
    }
}